src/main/java/dev/openfeature/sdk/EventEmitter.java

```diff
@@ -0,0 +1,108 @@
+package dev.openfeature.sdk;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import dev.openfeature.sdk.internal.AutoCloseableLock;
+import dev.openfeature.sdk.internal.AutoCloseableReentrantReadWriteLock;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Event emitter construct to be used by providers.
+ */
+@Slf4j
+public class EventEmitter {
+
+    private static final ExecutorService taskExecutor = Executors.newCachedThreadPool();
+    private final Map<ProviderEvent, List<Consumer<EventDetails>>> handlerMap;
+    private AutoCloseableReentrantReadWriteLock handlersLock = new AutoCloseableReentrantReadWriteLock();
+
+    /**
+     * Construct a new EventEmitter.
+     */
+    public EventEmitter() {
+        handlerMap = new ConcurrentHashMap<ProviderEvent, List<Consumer<EventDetails>>>();
+        handlerMap.put(ProviderEvent.PROVIDER_READY, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_CONFIGURATION_CHANGED, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_ERROR, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_STALE, new ArrayList<>());
+    }
+
+    /**
+     * Emit an event.
+     *
+     * @param event   Event type to emit.
+     * @param details Event details.
+     */
+    public void emit(ProviderEvent event, ProviderEventDetails details) {
+        try (AutoCloseableLock __ = handlersLock.readLockAutoCloseable()) {
+            EventDetails eventDetails = EventDetails.builder()
+                    .eventMetadata(details.getEventMetadata())
+                    .flagsChanged(details.getFlagsChanged())
+                    .message(details.getMessage())
+                    .build();
+
+            // we may be forwarding this event, preserve the name if so.
+            if (EventDetails.class.isInstance(details)) {
```

Comment:
```suggestion
             if (details instanceof EventDetails) {
```

IMO, instanceof is easy to read and understand

Reply:
agreed.